src/components/engagement/handlers/update-project-status.handler.ts

```diff
@@ -1,47 +1,97 @@
 import { node, relation } from 'cypher-query-builder';
+import { Session } from '../../../common';
 import { DatabaseService, EventsHandler, IEventHandler } from '../../../core';
-import { ProjectStatus, ProjectType } from '../../project';
+import {
+  ProjectRules,
+  ProjectStatus,
+  ProjectStep,
+  ProjectType,
+} from '../../project';
 import { ProjectUpdatedEvent } from '../../project/events';
-import { EngagementStatus } from '../dto';
+import { EngagementStatus, TerminalEngagementStatuses } from '../dto';
 import { EngagementService } from '../engagement.service';
 
 @EventsHandler(ProjectUpdatedEvent)
 export class UpdateProjectStatusHandler
   implements IEventHandler<ProjectUpdatedEvent> {
   constructor(
     private readonly db: DatabaseService,
-    private readonly engagementService: EngagementService
+    private readonly engagementService: EngagementService,
+    private readonly projectRules: ProjectRules
   ) {}
 
   async handle({ previous, updated, updates, session }: ProjectUpdatedEvent) {
-    // When the project becomes Active if it is currently InDevelopment
+    // every project step that triggers engagement status updates with accompanying engagement status
+    const updateStatusTuples: Array<[ProjectStep, EngagementStatus]> = [
+      [ProjectStep.Terminated, EngagementStatus.Terminated],
+      [ProjectStep.DidNotDevelop, EngagementStatus.DidNotDevelop],
+      [ProjectStep.Rejected, EngagementStatus.Rejected],
+      [ProjectStep.Completed, EngagementStatus.Completed],
+      [ProjectStep.Active, EngagementStatus.Active],
+    ];
+    // if the previous project step is the same as the updated step, we do nothing to the engagements
+    const [projectStep, engagementStatus] = updateStatusTuples.find(
+      (t) => t[0] === updated.step.value && t[0] !== previous.step.value
+    ) ?? [null, null];
+    if (!engagementStatus) return;
+    // we don't allow completing a project if there are non terminal engagements unless the user can bypass workflow constraints
+    // in which case update all non terminal engagements to Completed
     if (
-      previous.status !== ProjectStatus.InDevelopment ||
-      updated.status !== ProjectStatus.Active
+      projectStep === ProjectStep.Completed &&
+      !(await this.projectRules.canBypassWorkflow(session))
```

Comment:
I thought we said this check wasn't actually needed?

Reply:
If it's not accessible via the FE I guess not. If a non admin updated via playground it would have undesired consequences. Probably everyone using playground will only be making updates as an admin.

Reply:
No we should not assume that if possible. You need check for enabled [here](https://github.com/SeedCompany/cord-api-v3/blob/1a650c80461fa19f5cfa6f4356c933c7ee557304/src/components/project/project.rules.ts#L835), which should prevent us from getting here if we aren't supposed to.